### DISCUSSION TITLE
Running from an R Shiny Server?

### DISCUSSION THREAD

**Discussion by jamesfuller-cdc** [EU-ECDC/epitweetr#67] [3. Q&A]
Hi ECDC colleagues,

Thanks as always for your great work in epidemic intelligence. We're interested in exploring this tool a little further. We have an R Shiny server available (RStudio Connect). Any tips on getting the epitweetr shiny app published to an R shiny server?

The usual process of running these two lines of code will not work, since the script actually needs to be formatted as a shiny app.

`library(epitweetr)`
`epitweetr_app("data_dir")`

Thanks,

James

**Comment by edelgado-ecdc** [comment]
Dear James,

Packaged applications can be deployed to RStudio Connect, but the process is a bit more manual. Since the RStudio IDE does not detect the usual structure of a ShinyApp, it may not show the Publish button on the top right corner as usual.
You can, however, run the publish instruction manually using the `rsconnect` package.  Provided there's a server and an account already configured, you can proceed as follows:

1.- Save the script containing the library import and the epitweetr_app call as 'app.R'.
2.- Make sure you have an account and a server already registered - this will be done if you have deployed other apps or documents in the past. Otherwise, you can configure this on the Global options of RStudio IDE.
3.- Make sure your current working directory is the same as the saved script.
4.- Run the following command: `deployApp(appDir = getwd(), appFiles = list('app.R'), appName = 'EpiTweetr', account = 'name_of\\your_account', server = 'url.of.your.server', upload = T)`. Feel free to include in appFiles all the scripts, files and resources you may additionally need.

Bear in mind that you may or may not be successful when deploying this way depending on your RStudio Connect server configuration - you may need to tweak it a bit before.

Hope it helps!

Kind regards,
Enrique

**Reply by forchard** [reply]
Dear James and Enrique,
This could be a nice improvement to epitweetr but it would need to make some changes on the tool.

Firstly, the epitweetr_app function does not return the app so it should be slightly modified in order to support shiny server deployment.

Secondly but more complex, epitweetr app needs both local access its data folder and its REST API to displaying indexed timeseries. There are also two R background tasks for collecting tweets and calculating alerts. 

A simple solution could be to have single machine runninng both epitweetr and shiny server but this may not be optimal. Better options can be studied to make epitweetr able to split roles on separated servers. But it wold require to improve epitweetr architecture.

Another approach could be to have a read only instance of epitweetr to run on the shiny server. In this case the server would need to be able to run the REST API epitweeetr::fs_loop('data_directory')

Happy to keep discussing on this, 

Best,
Francisco

**Comment by jamesfuller-cdc** [comment]
Thanks @edelgado-ecdc  and @forchard . We also have an R Studio Server that can run scripts/background processes and store data. We'll continue exploring some options.
